src/v/archival/archiver_operations_impl.h

```diff
@@ -0,0 +1,211 @@
+
+    virtual model::initial_revision_id get_initial_revision() const = 0;
+
+    // aborted transactions
+    virtual ss::future<fragmented_vector<model::tx_range>>
+    aborted_transactions(model::offset base, model::offset last) const = 0;
+
+    // Replicate archival metadata add_segment command
+    //
+    // The method accepts 'clean_offset' which is the insync offset of the
+    // last uploaded manifest. This is a projected clean offset. It is possible
+    // for the archiver to upload the manifest  in parallel with segments.
+    // This means that when the `add_segments` command is called the uploaded
+    // version of the manifest becomes out of sync. The manifest upload is going
+    // one step behind the replication. Instead of uploading manifest after
+    // replication we're doing this before the replication to pipeline
+    // operations.
+    //
+    // The data upload workflow keeps track of this by storing two offsets.
+    // - clean offset is the offset of the last uploaded manifest
+    // - dirty offset is the current offset of the STM manifest
+    //
+    // If clean != dirty then the manifest in the cloud doesn't match the STM.
+    // The 'clean' offset is passed into this method and triggers 'mark_clean'
+    // command to be added to the replicated batch. When the commands are
+    // applied to the STM the 'dirty' offset is propagated forward. This offset
+    // is returned from this method and passed back to the workflow so the
+    // workflow could track this offset and inform the scheduler about the out
+    // of sync manifest.
+    virtual ss::future<result<model::offset>> add_segments(
+      std::vector<cloud_storage::segment_meta>,
+      std::optional<model::offset> clean_offset,
+      std::optional<model::offset> read_write_fence,
+      model::producer_id highest_pid,
+      ss::lowres_clock::time_point deadline,
+      ss::abort_source& external_as,
+      bool is_validated) noexcept
+      = 0;
+
+    virtual cloud_storage::remote_segment_path
+    get_remote_segment_path(const cloud_storage::segment_meta&)
+      = 0;
+
+    virtual cloud_storage::remote_manifest_path
+    get_remote_manifest_path(const cloud_storage::base_manifest&)
+      = 0;
+};
+
+struct cluster_partition_manager_api {
+    cluster_partition_manager_api() = default;
+    virtual ~cluster_partition_manager_api() = default;
+    cluster_partition_manager_api(const cluster_partition_manager_api&)
+      = delete;
+    cluster_partition_manager_api(cluster_partition_manager_api&&) = delete;
+    cluster_partition_manager_api&
+    operator=(const cluster_partition_manager_api&)
+      = delete;
+    cluster_partition_manager_api& operator=(cluster_partition_manager_api&&)
+      = delete;
+
+    virtual ss::shared_ptr<cluster_partition_api>
+    get_partition(const model::ntp&) = 0;
+};
+
+using upload_result = cloud_storage::upload_result;
+
+/// Wrapper for the cloud_storage::remote
+struct cloud_storage_remote_api {
```

Comment:
Is `remote` going to inherit from here in the final version? I added something similar for inventory service testing in https://github.com/redpanda-data/redpanda/blob/dev/src/v/cloud_storage/remote.h#L125-L149, 

I don't think there will be a resolution conflict or a problem as the inventory service and archiver use different methods of remote, but something to keep in mind.